packages/platforms/accton/arm64/as456x-52p/src/modules/arm64-accton-as456x-52p-mux.c

```diff
@@ -0,0 +1,233 @@
+	{}
+};
+MODULE_DEVICE_TABLE(of, as456x_mux_of_match);
+
+/* Write to mux register. Don't use i2c_transfer()/i2c_smbus_xfer()
+   for this as they will try to lock adapter a second time */
+static int as456x_mux_write(struct i2c_adapter *adap,
+				 struct i2c_client *client, u8 reg, u8 val)
+{
+	union i2c_smbus_data data;
+
+	data.byte = val;
+	return __i2c_smbus_xfer(adap, client->addr, client->flags,
+				I2C_SMBUS_WRITE, reg, I2C_SMBUS_BYTE_DATA, &data);
+}
+
+static int as456x_mux_select_chan(struct i2c_mux_core *muxc, u32 chan)
+{
+	struct as456x_mux_data *data = i2c_mux_priv(muxc);
+	struct i2c_client *client = data->client;
+	int ret = 0;
+
+	mutex_lock(&data->update_lock);
+	switch (data->type) {
+	case as456x_cpldm_mux:
+		ret = as456x_mux_write(muxc->parent, client,
+				CPLD_CHANNEL_SELECT_REG, 1 << chan);
+		break;
+	case pca9641_mux:
+	{
+		int i = 0;
+		u8 regs[] = { 0x5, 0x3, 0x1, 0x1 };
+		u8 vals[] = { 0x7b, 0x0, 0x1, 0x5 };
+
+		for (i = 0; i < ARRAY_SIZE(regs); i++) {
+			ret = as456x_mux_write(muxc->parent, client, regs[i], vals[i]);
+			if (unlikely(ret < 0)) {
+				ret = -EIO;
+
+				/* Try to close the mux channel */
+				as456x_mux_write(muxc->parent, client, 0x1, 0);
+				break;
+			}
+		}
+		break;
+	}
+	default:
+		break;
+	}
+
+	mutex_unlock(&data->update_lock);
+	return ret;
+}
+
+static int as456x_mux_deselect_mux(struct i2c_mux_core *muxc, u32 chan)
+{
+	struct as456x_mux_data *data = i2c_mux_priv(muxc);
+	struct i2c_client *client = data->client;
+	int ret = 0;
+
+	mutex_lock(&data->update_lock);
+
+	switch (data->type) {
+	case as456x_cpldm_mux:
+		ret = as456x_mux_write(muxc->parent, client,
+				CPLD_CHANNEL_SELECT_REG, CPLD_CHANNEL_DESELECT_VAL);
+		break;
+	case pca9641_mux:
+		ret = as456x_mux_write(muxc->parent, client, 0x1, 0);
+		break;
+	default:
+		break;
+	}
+
+	mutex_unlock(&data->update_lock);
+	return ret;
+}
+
+static void as456x_mux_cleanup(struct i2c_mux_core *muxc)
+{
+	i2c_mux_del_adapters(muxc);
+}
+
+/*
+ * I2C init/probing/exit functions
+ */
+static int as456x_mux_probe(struct i2c_client *client,
+			 const struct i2c_device_id *id)
+{
+	struct i2c_adapter *adap = to_i2c_adapter(client->dev.parent);
+	struct device *dev = &client->dev;
+	struct as456x_mux_data *data;
+	struct i2c_mux_core *muxc;
+	int ret = -ENODEV;
+	int i = 0;
+
+	if (!i2c_check_functionality(adap, I2C_FUNC_SMBUS_BYTE))
+		return -ENODEV;
+
+	muxc = i2c_mux_alloc(adap, dev, CPLD_MAX_NCHANS, sizeof(*data), 0,
+				as456x_mux_select_chan, as456x_mux_deselect_mux);
+	if (!muxc)
+		return -ENOMEM;
+
+	data = i2c_mux_priv(muxc);
+	mutex_init(&data->update_lock);
+	data->type = id->driver_data;
+	data->client = client;
+	i2c_set_clientdata(client, muxc);
+
+	/* Now create an adapter for each channel */
+	for (i = 0; i < chips[data->type].nchans; i++) {
+		ret = i2c_mux_add_adapter(muxc, 0, i, 0);
+		if (ret)
+			goto exit_mux;
+	}
+
+	return 0;
+
+exit_mux:
+	as456x_mux_cleanup(muxc);
```

Comment:
To avoid a goto here:

```suggestion
		if (ret)
			break;
	}

	if (ret)
		as456x_mux_cleanup(muxc);
```

Reply:
Modified.